pkg/canary/knative_controller.go

```diff
@@ -0,0 +1,161 @@
+	return setStatusWeight(kc.flaggerClient, cd, val)
+}
+
+// SetStatusIterations updates the canary status iterations value
+func (kc *KnativeController) SetStatusIterations(cd *flaggerv1.Canary, val int) error {
+	return setStatusIterations(kc.flaggerClient, cd, val)
+}
+
+// SetStatusPhase updates the canary status phase
+func (kc *KnativeController) SetStatusPhase(cd *flaggerv1.Canary, phase flaggerv1.CanaryPhase) error {
+	return setStatusPhase(kc.flaggerClient, cd, phase)
+}
+
+func (kc *KnativeController) Initialize(cd *flaggerv1.Canary) (bool, error) {
+	if cd.Status.Phase == "" || cd.Status.Phase == flaggerv1.CanaryPhaseInitializing {
+		service, err := kc.knativeClient.ServingV1().Services(cd.Namespace).Get(context.TODO(), cd.Spec.TargetRef.Name, metav1.GetOptions{})
+		if err != nil {
+			return true, fmt.Errorf("service %s.%s get query error: %w", cd.Spec.TargetRef.Name, cd.Namespace, err)
+		}
+
+		service.Annotations["flagger.app/primary-revision"] = service.Status.LatestCreatedRevisionName
+		percent := int64(100)
+		traffic := make([]v1.TrafficTarget, 1)
+		traffic = append(traffic, v1.TrafficTarget{Percent: &percent, RevisionName: service.Status.LatestCreatedRevisionName})
+		service.Spec.Traffic = traffic
+
+		_, err = kc.knativeClient.ServingV1().Services(cd.Namespace).Update(context.TODO(), service, metav1.UpdateOptions{})
+		if err != nil {
+			return true, fmt.Errorf("service %s.%s update query error %w", cd.Spec.TargetRef.Name, cd.Namespace, err)
+		}
+	}
+
+	return true, nil
+}
+
+func (kc *KnativeController) Promote(cd *flaggerv1.Canary) error {
+	service, err := kc.knativeClient.ServingV1().Services(cd.Namespace).Get(context.TODO(), cd.Spec.TargetRef.Name, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("service %s.%s get query error: %w", cd.Spec.TargetRef.Name, cd.Namespace, err)
+	}
+
+	service.Annotations["flagger.app/primary-revision"] = service.Status.LatestCreatedRevisionName
+	_, err = kc.knativeClient.ServingV1().Services(cd.Namespace).Update(context.TODO(), service, metav1.UpdateOptions{})
+	if err != nil {
+		return fmt.Errorf("service %s.%s update query error %w", cd.Spec.TargetRef.Name, cd.Namespace, err)
+	}
+
+	return nil
+}
+
+func (kc *KnativeController) HasTargetChanged(cd *flaggerv1.Canary) (bool, error) {
+	service, err := kc.knativeClient.ServingV1().Services(cd.Namespace).Get(context.TODO(), cd.Spec.TargetRef.Name, metav1.GetOptions{})
+	if err != nil {
+		return true, fmt.Errorf("service %s.%s get query error: %w", cd.Spec.TargetRef.Name, cd.Namespace, err)
+	}
+
+	return hasSpecChanged(cd, service.Status.LatestCreatedRevisionName)
+}
+
+func (kc *KnativeController) HaveDependenciesChanged(canary *flaggerv1.Canary) (bool, error) {
+	// TODO: Not sure if we'd need this for Knative deployments
+	return false, nil
+}
+
+func (kc *KnativeController) ScaleToZero(canary *flaggerv1.Canary) error {
+	// Not Implemented: Not needed for Knative deployments
```

Comment:
Just curious what this does for other providers?

Reply:
In other providers flagger maintains two instances of the application. From memory the process looks like this:

1. When a new version is created update the canary & scale it up
2. Slowly send traffic from the primary to the canary & monitor
3. Once the canary reaches 100% & is performing as expected, update the primary
4. Switch 100% traffic over to the primary
5. Scale the canary back to zero